### ISSUE TITLE
Crashes Visual Studio 16.7.5

### ISSUE THREAD

**Issue by cryolithic** [sentryone/unittestgenerator#42]
Smaller C# project, I added a new generic collection, and upon hitting Generate Unit Tests, devenv.exe becomes unresponsive until it eventually crashes. I ran with /log, but activity log doesn't seem to show anything relevant.

VS Info:

`Microsoft Visual Studio Professional 2019
Version 16.7.5
VisualStudio.16.Release/16.7.5+30523.141
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

Visual C++ 2019   00435-60000-00000-AA930
Microsoft Visual C++ 2019

.NET Portability Analyzer   1.1.10808.0
Evaluates portability of assemblies across .NET platforms.

ASP.NET and Web Tools 2019   16.7.532.28833
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2035807+72d099b977d3a85e65fa3b0614ca8cfc803fef02
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.7.532.28833
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.7.532.28833
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.7.532.28833
Azure Functions and Web Jobs Tools

C# Tools   3.7.0-6.20459.4+7ee7c540e3bf19765695b1331c946ddf986d1516
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Cookiecutter   16.7.20248.6
Provides tools for finding, instantiating and customizing templates in cookiecutter format.

Entity Framework Core Power Tools   2.4
Adds useful design-time EF Core DbContext features to the Visual Studio Solution Explorer context menu. 

gmaster for Visual Studio   1.0.646
gmaster integration with Visual Studio

Google Test Adapter   1.0
Google Test Adapter adds support for the C++ test framework Google Test

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

JetBrains ReSharper 2020.2.4   Build 202.0.20200925.65451
JetBrains ReSharper package for Microsoft Visual Studio. For more information about ReSharper, visit http://www.jetbrains.com/resharper. Copyright © 2020 JetBrains, Inc.

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.30701.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.79+ge3567815aa.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   5.7.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Project System Tools   1.0
Tools for working with C#, VisualBasic, and F# projects.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Python   16.7.20248.6
Provides IntelliSense, projects, templates, debugging, interactive windows, and other support for Python developers.

Python - Conda support   16.7.20248.6
Conda support for Python projects.

Python - Django support   16.7.20248.6
Provides templates and integration for the Django web framework.

Python - IronPython support   16.7.20248.6
Provides templates and integration for IronPython-based projects.

Python - Profiling support   16.7.20248.6
Profiling support for Python projects.

SentryOne Unit Test Generator Extension   1.0
Generates unit test boiler plate for existing code

Sharpen   0.10.1
Sharpen is a Visual Studio extension that intelligently introduces new C# language features into your existing code base. It makes your code smaller, simpler, faster, more expressive and more readable, in one word - sharper.

SQL Server Data Tools   16.0.62008.03130
Microsoft SQL Server Data Tools

Syntax Visualizer   1.0
An extension for visualizing Roslyn SyntaxTrees.

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   16.0.20702.2001
TypeScript Tools for Microsoft Visual Studio

Viasfora   4.3.195
Add color to your Visual Studio editor!

Visual Basic Tools   3.7.0-6.20459.4+7ee7c540e3bf19765695b1331c946ddf986d1516
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual C++ for Cross Platform Mobile Development (Android)   16.0.30309.115
Visual C++ for Cross Platform Mobile Development (Android)

Visual C++ for Cross Platform Mobile Development (iOS)   16.0.30309.115
Visual C++ for Cross Platform Mobile Development (iOS)

Visual C++ for Linux Development   1.0.9.30523
Visual C++ for Linux Development

Visual F# Tools 10.10.0.0 for F# 4.7   16.7.0-beta.20361.3+3ef6f0b514198c0bfa6c2c09fefe41a740b024d5
Microsoft Visual F# Tools 10.10.0.0 for F# 4.7

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions (Preview)   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Kubernetes   1.0
Visual Studio Tools for Kubernetes`

Sysinfo

`OS Name	Microsoft Windows 10 Enterprise
Version	10.0.19041 Build 19041
Other OS Description 	Not Available
OS Manufacturer	Microsoft Corporation
System Manufacturer	Dell Inc.
System Model	Precision 5820 Tower X-Series
System Type	x64-based PC
Processor	Intel(R) Core(TM) i9-10980XE CPU @ 3.00GHz, 3000 Mhz, 18 Core(s), 36 Logical Processor(s)
BIOS Version/Date	Dell Inc. 2.4.0, 7/6/2020
SMBIOS Version	3.2
Embedded Controller Version	255.255
BIOS Mode	UEFI
Hardware Abstraction Layer	Version = "10.0.19041.488"
Installed Physical Memory (RAM)	128 GB
Total Physical Memory	128 GB
Available Physical Memory	92.7 GB
Total Virtual Memory	147 GB
Available Virtual Memory	105 GB
Page File Space	19.0 GB
Page File	C:\pagefile.sys
Kernel DMA Protection	Not Available
`

**Comment by mattwhitfield** [comment]
Thanks for the bug report!

If it's crashing the host process it's usually something like a stack overflow. I don't suppose you would be able to post a small project that reproduces the issue? If you're not comfortable posting it publically, I'd be happy to receive it personally if that's an option.

Thanks again, much appreciated

**Comment by cryolithic** [comment]
this class seems to do it. Specifically the non generic CopyTo. I tried a few different iterations and most other simple things I tried worked. This class will crash on test generation in multiple slns.

Removing the CopyTo from my code allowed me to generate with the rest of the actual class as well. Odd.

Additional edit: It generated, but it generated...weird. It referenced a private member of the test class in the constructor tests, but never made the member. Easy fix, but figured I'd mention it.

```using System;
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace sample
{
	public class DelayedConcurrentQueue<T> : IProducerConsumerCollection<T>
	{
		private readonly TimeSpan _delay;
		private readonly ConcurrentQueue<Tuple<DateTime, T>> _internalQueue;
		private int _count;
		private bool _isSynchronized;
		private object _syncRoot;

		public void CopyTo(Array destinationArray, int index)
		{
			if (destinationArray.Length - index <= _internalQueue.Count)
			{
				throw new
					ArgumentException("Destination destinationArray was not long enough. Check the destination index, length, and the destinationArray's lower bounds. (Parameter 'destinationArray')");
			}

			var source = _internalQueue.Select(t => t.Item2).ToArray();
			Array.Copy(source, destinationArray, index);
		}

		int ICollection.Count => _count;

		bool ICollection.IsSynchronized => _isSynchronized;

		object ICollection.SyncRoot => _syncRoot;

		IEnumerator<T> IEnumerable<T>.GetEnumerator()
		{
			throw new NotImplementedException();
		}

		IEnumerator IEnumerable.GetEnumerator()
		{
			throw new NotImplementedException();
		}

		void IProducerConsumerCollection<T>.CopyTo(T[] array, int index)
		{
			throw new NotImplementedException();
		}

		T[] IProducerConsumerCollection<T>.ToArray()
		{
			throw new NotImplementedException();
		}

		bool IProducerConsumerCollection<T>.TryAdd(T item)
		{
			throw new NotImplementedException();
		}

		bool IProducerConsumerCollection<T>.TryTake(out T item)
		{
			throw new NotImplementedException();
		}
	}
}

**Comment by mattwhitfield** [comment]
Thanks for the sample. I've opened a PR to fix the issue, and also add support for generating test stubs for explicitly implemented methods, which looks like it would be useful to you 👍

**Comment by mattwhitfield** [comment]
For this bit

> Additional edit: It generated, but it generated...weird. It referenced a private member of the test class in the constructor tests, but never made the member. Easy fix, but figured I'd mention it.

Can you open another issue if that persists? Given what I had to fix for this I think that might be a different thing.

**Comment by cryolithic** [comment]
I'll have a fair number more tests to generate in the next while, if it comes up again, I'll open a new one. If I have time, I'll debug and submit a PR, but when do programmers have spare time?

**Comment by mattwhitfield** [comment]
Just to confirm - this should be resolved by 1.0.35 which is now on the VS marketplace. Thanks again!
